File: web/src/app/features/treasury/treasury.service.ts
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';
import { ApiService } from '../../core/api/api.service';

export interface TreasuryPlacement {
  id: string;
  reference: string;
  counterpartyName: string;
  counterpartyBic?: string;
  placementType: 'FIXED_DEPOSIT' | 'TREASURY_BILL' | 'BOND' | 'CALL_MONEY' | 'REPO';
  principal: number;
  interestRate: number;
  currencyCode: string;
  startDate: string;
  maturityDate: string;
  status: 'PENDING' | 'ACTIVE' | 'MATURED' | 'CANCELLED';
  expectedReturn?: number;
  actualReturn?: number;
  glSourceAccount?: string;
  glIncomeAccount?: string;
  notes?: string;
  createdAt: string;
  updatedAt: string;
}

export interface TreasuryInterbankPosition {
  id: string;
  reference: string;
  counterpartyName: string;
  counterpartyBic?: string;
  direction: 'LENDING' | 'BORROWING';
  amount: number;
  currencyCode: string;
  interestRate: number;
  startDate: string;
  maturityDate?: string;
  status: 'ACTIVE' | 'SETTLED' | 'CANCELLED';
  settlementGl?: string;
  notes?: string;
  createdAt: string;
  updatedAt: string;
}

export interface PlacementRequest {
  reference: string;
  counterpartyName: string;
  counterpartyBic?: string;
  placementType: string;
  principal: number;
  interestRate: number;
  currencyCode: string;
  startDate: string;
  maturityDate: string;
  expectedReturn?: number;
  notes?: string;
}

export interface InterbankRequest {
  reference: string;
  counterpartyName: string;
  counterpartyBic?: string;
  direction: string;
  amount: number;
  currencyCode: string;
  interestRate: number;
  startDate: string;
  maturityDate?: string;
  notes?: string;
}

@Injectable({ providedIn: 'root' })
export class TreasuryService {
  private readonly api = inject(ApiService);

  // ── Placements ──────────────────────────────────────────────────────────────

  listPlacements(): Observable<TreasuryPlacement[]> {
    return this.api.get<TreasuryPlacement[]>('/treasury/placements');
  }

  getPlacement(id: string): Observable<TreasuryPlacement> {
    return this.api.get<TreasuryPlacement>(`/treasury/placements/${id}`);
  }

  createPlacement(req: PlacementRequest): Observable<TreasuryPlacement> {
    return this.api.post<TreasuryPlacement>('/treasury/placements', req);
  }

  updatePlacement(id: string, req: PlacementRequest): Observable<TreasuryPlacement> {
    return this.api.put<TreasuryPlacement>(`/treasury/placements/${id}`, req);
  }

  commandPlacement(id: string, command: string): Observable<TreasuryPlacement> {
    return this.api.command<TreasuryPlacement>(`/treasury/placements/${id}`, command);
  }

  deletePlacement(id: string): Observable<void> {
    return this.api.delete<void>(`/treasury/placements/${id}`);
  }

  // ── Interbank Positions ────────────────────────────────────────────────────

  listPositions(): Observable<TreasuryInterbankPosition[]> {
    return this.api.get<TreasuryInterbankPosition[]>('/treasury/positions');
  }

  getPosition(id: string): Observable<TreasuryInterbankPosition> {
    return this.api.get<TreasuryInterbankPosition>(`/treasury/positions/${id}`);
  }

  createPosition(req: InterbankRequest): Observable<TreasuryInterbankPosition> {
    return this.api.post<TreasuryInterbankPosition>('/treasury/positions', req);
  }

  updatePosition(id: string, req: InterbankRequest): Observable<TreasuryInterbankPosition> {
    return this.api.put<TreasuryInterbankPosition>(`/treasury/positions/${id}`, req);
  }

  commandPosition(id: string, command: string): Observable<TreasuryInterbankPosition> {
    return this.api.command<TreasuryInterbankPosition>(`/treasury/positions/${id}`, command);
  }

  deletePosition(id: string): Observable<void> {
    return this.api.delete<void>(`/treasury/positions/${id}`);
  }

  // ── Liquidity ────────────────────────────────────────────────────────────────

  getLiquidityPositions(): Observable<LiquidityPosition[]> {
    return this.api.get<LiquidityPosition[]>('/treasury/liquidity/positions');
  }

  getLiquidityPosition(currency: string): Observable<LiquidityPosition> {
    return this.api.get<LiquidityPosition>(`/treasury/liquidity/positions/${currency}`);
  }

  getCashFlowForecast(currency: string, days = 30): Observable<CashFlowEntry[]> {
    return this.api.get<CashFlowEntry[]>(`/treasury/liquidity/cashflow?currency=${currency}&days=${days}`);
  }

  listReserveRequirements(): Observable<ReserveRequirement[]> {
    return this.api.get<ReserveRequirement[]>('/treasury/liquidity/reserves');
  }

  createReserveRequirement(req: ReserveRequest): Observable<ReserveRequirement> {
    return this.api.post<ReserveRequirement>('/treasury/liquidity/reserves', req);
  }

  updateReserveRequirement(id: string, req: ReserveRequest): Observable<ReserveRequirement> {
    return this.api.put<ReserveRequirement>(`/treasury/liquidity/reserves/${id}`, req);
  }

  deleteReserveRequirement(id: string): Observable<void> {
    return this.api.delete<void>(`/treasury/liquidity/reserves/${id}`);
  }

  getLiquiditySnapshots(currency: string, limit = 30): Observable<LiquiditySnapshot[]> {
    return this.api.get<LiquiditySnapshot[]>(`/treasury/liquidity/snapshots?currency=${currency}&limit=${limit}`);
  }

  takeSnapshot(): Observable<void> {
    return this.api.post<void>('/treasury/liquidity/snapshots/take', {});
  }
}

// ── Liquidity interfaces ─────────────────────────────────────────────────────

export interface LiquidityPosition {
  currency: string;
  cashOnHand: number;
  placementsDeployed: number;
  interbankLending: number;
  interbankBorrowing: number;
  netLiquidityPosition: number;
  reserveRequirement: number;
  surplusDeficit: number;
  alertLevel: 'OK' | 'WARN' | 'BREACH';
  asOfDate: string;
}

export interface CashFlowEntry {
  date: string;
  type: string;
  reference: string;
  amount: number;
  currency: string;
  direction: 'INFLOW' | 'OUTFLOW';
}

export interface ReserveRequirement {
  id: string;
  currencyCode: string;
  minimumBalance: number;
  minimumRatioPercent?: number;
  alertThresholdPercent?: number;
  regulatoryReference?: string;
  active: boolean;
}

export interface ReserveRequest {
  currencyCode: string;
  minimumBalance: number;
  minimumRatioPercent?: number;
  alertThresholdPercent?: number;
  regulatoryReference?: string;
}

export interface LiquiditySnapshot {
  id: string;
  snapshotDate: string;
  currencyCode: string;
  cashOnHand: number;
  placementsDeployed: number;
  interbankLending: number;
  interbankBorrowing: number;
  netLiquidityPosition: number;
  reserveRequirement?: number;
  surplusDeficit?: number;
  createdAt: string;
}

File: backend/src/main/java/com/cba/treasury/TreasuryService.java
package com.cba.treasury;

import com.cba.common.exception.CbaException;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Locale;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class TreasuryService {

    private final TreasuryPlacementRepository placementRepo;
    private final TreasuryInterbankPositionRepository interbankRepo;

    // ── Placements ──────────────────────────────────────────────────────────────

    @Transactional(readOnly = true)
    public List<TreasuryPlacement> listPlacements() {
        return placementRepo.findAll();
    }

    @Transactional(readOnly = true)
    public TreasuryPlacement getPlacement(UUID id) {
        return placementRepo.findById(id)
                .orElseThrow(() -> CbaException.notFound("TreasuryPlacement", id));
    }

    @Transactional
    public TreasuryPlacement createPlacement(TreasuryPlacementRequest req) {
        if (placementRepo.findByReference(req.reference()).isPresent()) {
            throw CbaException.conflict("DUPLICATE_REFERENCE", "Reference " + req.reference() + " already exists");
        }
        var p = new TreasuryPlacement();
        applyPlacementRequest(p, req);
        return placementRepo.save(p);
    }

    @Transactional
    public TreasuryPlacement updatePlacement(UUID id, TreasuryPlacementRequest req) {
        var p = getPlacement(id);
        applyPlacementRequest(p, req);
        return placementRepo.save(p);
    }

    @Transactional
    public TreasuryPlacement commandPlacement(UUID id, String command) {
        var p = getPlacement(id);
        switch (command.toLowerCase(Locale.ROOT)) {
            case "activate" -> {
                if (p.getStatus() != TreasuryPlacement.Status.PENDING)
                    throw CbaException.badRequest("INVALID_STATE", "Only PENDING placements can be activated");
                p.setStatus(TreasuryPlacement.Status.ACTIVE);
            }
            case "mature" -> {
                if (p.getStatus() != TreasuryPlacement.Status.ACTIVE)
                    throw CbaException.badRequest("INVALID_STATE", "Only ACTIVE placements can be matured");
                p.setStatus(TreasuryPlacement.Status.MATURED);
            }
            case "cancel" -> {
                if (p.getStatus() == TreasuryPlacement.Status.MATURED)
                    throw CbaException.badRequest("INVALID_STATE", "Matured placements cannot be cancelled");
                p.setStatus(TreasuryPlacement.Status.CANCELLED);
            }
            default -> throw CbaException.badRequest("UNKNOWN_COMMAND", "Unknown command: " + command);
        }
        return placementRepo.save(p);
    }

    @Transactional
    public void deletePlacement(UUID id) {
        var p = getPlacement(id);
        if (p.getStatus() == TreasuryPlacement.Status.ACTIVE)
            throw CbaException.badRequest("INVALID_STATE", "Cannot delete an ACTIVE placement");
        placementRepo.delete(p);
    }

    // ── Interbank Positions ────────────────────────────────────────────────────

    @Transactional(readOnly = true)
    public List<TreasuryInterbankPosition> listPositions() {
        return interbankRepo.findAll();
    }

    @Transactional(readOnly = true)
    public TreasuryInterbankPosition getPosition(UUID id) {
        return interbankRepo.findById(id)
                .orElseThrow(() -> CbaException.notFound("TreasuryInterbankPosition", id));
    }

    @Transactional
    public TreasuryInterbankPosition createPosition(TreasuryInterbankRequest req) {
        if (interbankRepo.findByReference(req.reference()).isPresent()) {
            throw CbaException.conflict("DUPLICATE_REFERENCE", "Reference " + req.reference() + " already exists");
        }
        var pos = new TreasuryInterbankPosition();
        applyInterbankRequest(pos, req);
        return interbankRepo.save(pos);
    }

    @Transactional
    public TreasuryInterbankPosition updatePosition(UUID id, TreasuryInterbankRequest req) {
        var pos = getPosition(id);
        applyInterbankRequest(pos, req);
        return interbankRepo.save(pos);
    }

    @Transactional
    public TreasuryInterbankPosition commandPosition(UUID id, String command) {
        var pos = getPosition(id);
        switch (command.toLowerCase(Locale.ROOT)) {
            case "settle" -> {
                if (pos.getStatus() != TreasuryInterbankPosition.Status.ACTIVE)
                    throw CbaException.badRequest("INVALID_STATE", "Only ACTIVE positions can be settled");
                pos.setStatus(TreasuryInterbankPosition.Status.SETTLED);
            }
            case "cancel" -> {
                if (pos.getStatus() != TreasuryInterbankPosition.Status.ACTIVE)
                    throw CbaException.badRequest("INVALID_STATE", "Only ACTIVE positions can be cancelled");
                pos.setStatus(TreasuryInterbankPosition.Status.CANCELLED);
            }
            default -> throw CbaException.badRequest("UNKNOWN_COMMAND", "Unknown command: " + command);
        }
        return interbankRepo.save(pos);
    }

    @Transactional
    public void deletePosition(UUID id) {
        var pos = getPosition(id);
        if (pos.getStatus() == TreasuryInterbankPosition.Status.ACTIVE)
            throw CbaException.badRequest("INVALID_STATE", "Cannot delete an ACTIVE position");
        interbankRepo.delete(pos);
    }

    // ── Helpers ────────────────────────────────────────────────────────────────

    private void applyPlacementRequest(TreasuryPlacement p, TreasuryPlacementRequest req) {
        p.setReference(req.reference());
        p.setCounterpartyName(req.counterpartyName());
        p.setCounterpartyBic(req.counterpartyBic());
        p.setPlacementType(TreasuryPlacement.PlacementType.valueOf(req.placementType()));
        p.setPrincipal(req.principal());
        p.setInterestRate(req.interestRate());
        p.setCurrencyCode(req.currencyCode() != null ? req.currencyCode() : "USD");
        p.setStartDate(req.startDate());
        p.setMaturityDate(req.maturityDate());
        p.setExpectedReturn(req.expectedReturn());
        p.setGlSourceAccount(req.glSourceAccount());
        p.setGlIncomeAccount(req.glIncomeAccount());
        p.setNotes(req.notes());
    }

    private void applyInterbankRequest(TreasuryInterbankPosition pos, TreasuryInterbankRequest req) {
        pos.setReference(req.reference());
        pos.setCounterpartyName(req.counterpartyName());
        pos.setCounterpartyBic(req.counterpartyBic());
        pos.setDirection(TreasuryInterbankPosition.Direction.valueOf(req.direction()));
        pos.setAmount(req.amount());
        pos.setCurrencyCode(req.currencyCode() != null ? req.currencyCode() : "USD");
        pos.setInterestRate(req.interestRate());
        pos.setStartDate(req.startDate());
        pos.setMaturityDate(req.maturityDate());
        pos.setSettlementGl(req.settlementGl());
        pos.setNotes(req.notes());
    }
}
